File: Leet Code Solutions/1286-constrained-subsequence-sum/1286-constrained-subsequence-sum.go
func constrainedSubsetSum(nums []int, k int) int {
    var dq []int
    for i := range nums {
        if len(dq) != 0 {
            nums[i] += nums[dq[0]]
        }
        
        for len(dq) != 0 && (i - dq[0] >= k || nums[i] >= nums[dq[len(dq) - 1]]) {
            if nums[i] >= nums[dq[len(dq) - 1]] {
                dq = dq[:len(dq)-1]
            } else {
                dq = dq[1:]
            }
        }

        if nums[i] > 0 {
            dq = append(dq, i)
        }
    }
    return max(nums)
}

func max(arr []int) int {
    maxVal := arr[0]
    for _, val := range arr {
        if val > maxVal {
            maxVal = val
        }
    }
    return maxVal
}

File: Leet Code Solutions/1286-constrained-subsequence-sum/1286-constrained-subsequence-sum.py
class Solution:
    def constrainedSubsetSum(self, nums: List[int], k: int) -> int:
        dq = deque()
        for i in range(len(nums)):
            nums[i] += nums[dq[0]] if dq else 0
            
            while dq and (i - dq[0] >= k or nums[i] >= nums[dq[-1]]):
                dq.pop() if nums[i] >= nums[dq[-1]] else dq.popleft()
                
            if nums[i] > 0:
                dq.append(i)
                
        return max(nums)